src/r/recordFactories/ProjectFactory.ts

```diff
@@ -419,6 +445,78 @@ export class ProjectFactory extends RecordFactory<RT.project> {
     }
   }
 
+  addLinkedVariables<N extends RT>(this: ProjectFactory, records: RecordNode<N>[]): void {
+    records.forEach(record => {
+      if (record?.type === RT.element) {
+        switch((record as RecordNode<RT.element>).props.element_type) {
+          case ElementType.media_upload: {
+            const variableRecord = this.addVariableOfType(VariableType.string);
+            variableRecord.props.var_category = VarCategory.autogenerated;
+            variableRecord.props.var_track = true;
+            variableRecord.props.var_default = "";
+            //Keep the name of the variable same as the lead gen field
+            //Going via changeRecordName to take care of duplicate names
+            this.changeRecordName(RT.variable, variableRecord.id, varNameFromOriginName(record.name));
+            //Link lead gen field to the variable id
+            (record as RecordNode<RT.element>).props.media_upload_var_id = variableRecord.id;
+            break;
+          }
+
+          case ElementType.embed_scorm: {
+            const scoreVariableRecord = this.addVariableOfType(VariableType.number);
+            scoreVariableRecord.props.var_category = VarCategory.autogenerated;
+            scoreVariableRecord.props.var_track = true;
+            scoreVariableRecord.props.var_default = 0;
+            this.changeRecordName(RT.variable, scoreVariableRecord.id, varNameFromOriginName(record.name));
+            (record as RecordNode<RT.element>).props.embed_scorm_score_var_id = scoreVariableRecord.id;
+
+            const suspendDataVariableRecord = this.addVariableOfType(VariableType.string);
+            suspendDataVariableRecord.props.var_category = VarCategory.autogenerated;
+            suspendDataVariableRecord.props.var_track = true;
+            suspendDataVariableRecord.props.var_default = 0;
+            this.changeRecordName(RT.variable, suspendDataVariableRecord.id, varNameFromOriginName(record.name));
+            (record as RecordNode<RT.element>).props.embed_scorm_suspend_data_var_id = suspendDataVariableRecord.id;
+
+            const progressVariableRecord = this.addVariableOfType(VariableType.number);
+            progressVariableRecord.props.var_category = VarCategory.autogenerated;
+            progressVariableRecord.props.var_track = true;
+            progressVariableRecord.props.var_default = 0;
+            this.changeRecordName(RT.variable, progressVariableRecord.id, varNameFromOriginName(record.name));
+            (record as RecordNode<RT.element>).props.embed_scorm_progress_var_id = progressVariableRecord.id;
+            break;
+          }
+        }
+      }
+    })
+  }
+
+  deleteLinkedVariables<N extends RT>(this: ProjectFactory, records: RecordNode<N>[]): void {
+    records.forEach(record => {
```

Comment:
Use for-of loop. Has better performance as it is called sync.
forEach is called in async fashion with a callback and incurs a function call overhead.